runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java

```diff
@@ -0,0 +1,223 @@
+        }
+
+        Vertex sourceVertex = edgeStartPoints.get(edgeId);
+        if (sourceVertex == null) {
+          throw new RuntimeException("Oops!");
+        }
+
+        Coder edgeCoder = edgeCoders.get(edgeId);
+        if (edgeCoder == null) {
+          throw new RuntimeException("Oops!");
+        }
+
+        List<Vertex> destinationVertices =
+            edgeEndPoints.getOrDefault(edgeId, Collections.emptyList());
+        boolean sideInputEdge = edgeId.contains("PCollectionView"); // todo: this is a hack!
+        for (Vertex destinationVertex : destinationVertices) {
+          addEdge(sourceVertex, destinationVertex, edgeCoder, edgeId, pCollId, sideInputEdge);
+        }
+      }
+    }
+
+    private void addEdge(
+        Vertex sourceVertex,
+        Vertex destinationVertex,
+        Coder coder,
+        String edgeId,
+        String pCollId,
+        boolean sideInputEdge) {
+      // todo: set up the edges properly, including other aspects too, like parallelism
+
+      try {
+        Edge edge =
+            Edge.from(sourceVertex, getNextFreeOrdinal(sourceVertex, false))
+                .to(destinationVertex, getNextFreeOrdinal(destinationVertex, true))
+                .distributed(); // todo: why is it always distributed?
+        if (sideInputEdge) {
+          edge = edge.broadcast();
+        } else {
+          edge =
+              edge.partitioned(
+                  new PartitionedKeyExtractor(
+                      coder)); // todo: we likely don't need to partition everything
+        }
+        dag.edge(edge);
+
+        String sourceVertexName = sourceVertex.getName();
+        String destinationVertexName = destinationVertex.getName();
+        for (WiringListener listener : listeners) {
+          listener.isInboundEdgeOfVertex(edge, edgeId, pCollId, destinationVertexName);
+          listener.isOutboundEdgeOfVertex(edge, edgeId, pCollId, sourceVertexName);
+        }
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    private int getNextFreeOrdinal(Vertex vertex, boolean inbound) {
+      Map<Vertex, Integer> ordinals = inbound ? inboundOrdinals : outboundOrdinals;
+      int nextOrdinal = 1 + ordinals.getOrDefault(vertex, -1);
+      ordinals.put(vertex, nextOrdinal);
+      return nextOrdinal;
+    }
+  }
+
+  private static class PartitionedKeyExtractor implements FunctionEx<byte[], Object> {
+    private final Coder coder;
+
+    PartitionedKeyExtractor(Coder coder) {
+      this.coder = coder;
+    }
+
+    @Override
+    public Object applyEx(byte[] b) throws Exception {
+      Object t = coder.decode(new ByteArrayInputStream(b)); // todo: decoding twice....
+      Object key = null;
+      if (t instanceof WindowedValue) {
+        t = ((WindowedValue) t).getValue();
+      }
+      if (t instanceof KV) {
+        key = ((KV) t).getKey();
+      }
+      return key == null ? "all" : key; // todo: why "all"?
```

Comment:
I would also be curious :)

Reply:
This "all" partitioning key is needed right now because we use partitioned edges all the time. But we shouldn't need to. I can't fix it immediately, but will look into it for the next PR.